File: runtime/app/schemas/feedback.py
"""FeedbackEvent schema — matches docs/tyndale-spec/L05_feedback_consent_schema.md."""

from __future__ import annotations

import datetime
from typing import Literal

from pydantic import BaseModel, Field

FeedbackType = Literal[
    "thumbs",
    "structured_correction",
    "outcome_report",
    "value_confirmation",
    "implicit_signal",
]

StructuredReason = Literal[
    "wrong_number",
    "missed_an_error",
    "false_error",
    "bad_recommendation",
    "confusing",
    "wrong_citation",
    "wrong_coverage_reading",
    "other",
]


class FeedbackOutcome(BaseModel):
    acted_on_recommendation: bool | None = None
    resolved: Literal["yes", "no", "partial", "pending", "unknown"] | None = None
    amount_saved: float | None = None
    outcome_notes: str | None = None


class ValueConfirmation(BaseModel):
    # confirmation_kind distinguishes a generic extracted-value confirmation
    # from an encounter line-item confirmation (Phase 2I / L07 step 4).
    confirmation_kind: Literal["extracted_value", "encounter_lineitem"] | None = None
    field: str | None = None
    tyndale_extracted: str | None = None
    user_corrected: str | None = None
    was_correct: bool | None = None


class FeedbackEventIn(BaseModel):
    # Required (per L05 capture_schema.json)
    event_id: str
    timestamp: datetime.datetime
    case_file_id: str
    feedback_type: FeedbackType
    # Optional
    user_id: str | None = None
    response_id: str | None = Field(default=None, description="Which Tyndale response this is about")
    thumbs: Literal["up", "down"] | None = None
    structured_reason: list[StructuredReason] | None = None
    free_text: str | None = None
    outcome: FeedbackOutcome | None = None
    # What happened on ONE call the user made, tapped in call mode (H6). Typed rather than
    # smuggled through outcome_notes so nothing has to sniff a string for meaning (DL-39), and
    # deliberately NOT an `outcome`: none of the three routes resolves a case, so this must
    # never reach the recovered tally or permanently retire the outcome follow-up.
    call_outcome: Literal["fixing_it", "pushed_back", "left_message"] | None = None
    value_confirmation: ValueConfirmation | None = None
    # improvement_consent on the request body is IGNORED by the server — the
    # POST /v1/feedback route reads the user's current consent from the users
    # table at insert time (never trust the client's claimed value).
    improvement_consent: bool = False
    promoted_to_eval: bool = False
    linked_golden_example_id: str | None = None


# --- Phase 2J response/query shapes -----------------------------------------
class FeedbackAck(BaseModel):
    event_id: str
    feedback_event_id: str
    queued_for_deid: bool


class FeedbackEventOut(BaseModel):
    """Projection of a stored feedback_event for the audit-results screen to
    restore per-target rating state across navigation."""

    feedback_event_id: str
    feedback_type: str
    response_id: str | None = None
    thumbs: Literal["up", "down"] | None = None
    structured_reason: list[str] | None = None
    created_at: str


class CaseFeedbackPayload(BaseModel):
    case_file_id: str
    events: list[FeedbackEventOut] = Field(default_factory=list)


class OutcomePrompt(BaseModel):
    case_file_id: str
    days_since_recommendation: int
    finding_summary: str = Field(description="e.g. 'Cost sharing miscalculation with UnitedHealthcare'")


class OutcomePromptsPayload(BaseModel):
    prompts: list[OutcomePrompt] = Field(default_factory=list)


class UserProfile(BaseModel):
    id: str
    first_name: str
    email: str
    user_type: Literal["user", "admin"]
    improvement_consent: bool
    created_at: str


class ConsentUpdate(BaseModel):
    improvement_consent: bool


class ConsentHistoryEntry(BaseModel):
    changed_at: str
    from_consent: bool
    to_consent: bool


class ConsentHistoryPayload(BaseModel):
    user_id: str
    history: list[ConsentHistoryEntry] = Field(default_factory=list)

File: packages/shared/src/feedback.ts
/**
 * Feedback capture API contracts (Phase 2J / L05).
 * Mirrors runtime/app/schemas/feedback.py.
 *
 * NOTE (DL-38): this module MUST be re-exported from src/index.ts. Phase 2H
 * silently dropped the dashboard re-export; Phase 2I caught it only because
 * typecheck was run locally. Keep the barrel export in sync.
 */

export type FeedbackType =
  | 'thumbs'
  | 'structured_correction'
  | 'outcome_report'
  | 'value_confirmation'
  | 'implicit_signal';

export type StructuredReason =
  | 'wrong_number'
  | 'missed_an_error'
  | 'false_error'
  | 'bad_recommendation'
  | 'confusing'
  | 'wrong_citation'
  | 'wrong_coverage_reading'
  | 'other';

export type ThumbsValue = 'up' | 'down';
export type ResolvedValue = 'yes' | 'no' | 'partial' | 'pending' | 'unknown';

/** Human-readable labels for the structured-reason chips (UI order matters). */
export const STRUCTURED_REASON_LABELS: { value: StructuredReason; label: string }[] = [
  { value: 'wrong_number', label: 'Wrong number' },
  { value: 'missed_an_error', label: 'Missed an error' },
  { value: 'false_error', label: "Flagged something that wasn't actually wrong" },
  { value: 'bad_recommendation', label: 'Bad recommendation' },
  { value: 'confusing', label: 'Confusing' },
  { value: 'wrong_citation', label: 'Wrong citation' },
  { value: 'wrong_coverage_reading', label: 'Misread my coverage' },
  { value: 'other', label: 'Something else' },
];

export interface FeedbackOutcome {
  acted_on_recommendation?: boolean | null;
  resolved?: ResolvedValue | null;
  amount_saved?: number | null;
  outcome_notes?: string | null;
}

export interface ValueConfirmation {
  confirmation_kind?: 'extracted_value' | 'encounter_lineitem' | null;
  field?: string | null;
  tyndale_extracted?: string | null;
  user_corrected?: string | null;
  was_correct?: boolean | null;
}

/** The POST /v1/feedback request body (matches FeedbackEventIn). */
export interface FeedbackEvent {
  event_id: string;
  timestamp: string; // ISO-8601
  case_file_id: string;
  feedback_type: FeedbackType;
  response_id?: string | null;
  thumbs?: ThumbsValue | null;
  structured_reason?: StructuredReason[] | null;
  free_text?: string | null;
  outcome?: FeedbackOutcome | null;
  value_confirmation?: ValueConfirmation | null;
}

export interface FeedbackAck {
  event_id: string;
  feedback_event_id: string;
  queued_for_deid: boolean;
}

export interface FeedbackEventOut {
  feedback_event_id: string;
  feedback_type: string;
  response_id: string | null;
  thumbs: ThumbsValue | null;
  structured_reason: string[] | null;
  created_at: string;
}

export interface CaseFeedbackPayload {
  case_file_id: string;
  events: FeedbackEventOut[];
}

export interface OutcomePrompt {
  case_file_id: string;
  days_since_recommendation: number;
  finding_summary: string;
}

export interface OutcomePromptsPayload {
  prompts: OutcomePrompt[];
}

export interface UserProfile {
  id: string;
  first_name: string;
  email: string;
  user_type: 'user' | 'admin';
  improvement_consent: boolean;
  created_at: string;
}

export interface ConsentHistoryEntry {
  changed_at: string;
  from_consent: boolean;
  to_consent: boolean;
}

export interface ConsentHistoryPayload {
  user_id: string;
  history: ConsentHistoryEntry[];
}
